### DISCUSSION TITLE
Custom Zones for Snapping

### DISCUSSION THREAD

**Discussion by DenizUgur** [rxhanson/Rectangle#826] [Ideas]
I've looked over the discussions but couldn't find a similar topic. Since the current code base is ready for something like this. Maybe we could give users an option to provide custom snap zones and/or select multiple zones and move the window to the bounding box of those selections.

I'm actually describing [FancyZones](https://docs.microsoft.com/en-us/windows/powertoys/fancyzones). I would like to contribute this idea to the Rectangle but I would need some guidelines for to succeed.

I have never developed anything in Swift but I'm a quick learner. The file in question is [SnappingManager.swift](https://github.com/rxhanson/Rectangle/blob/master/Rectangle/Snapping/SnappingManager.swift)

In my mind, a rough idea for implementation could be like this. POC of course:

1. Read the configuration from a file or preferences. The option might describe 12x12 grid and each zone will be an area within this grid.
2. I see that you are defining some areas at Line 37. Maybe parse the configuration here. It would be enough to just snap to a location.
3. For the other option (multiple zones and bounding box of it), while left click + shift is pressed (or just right click) continue adding zones on to an array and find the bounding box of that array when the key combination is released. All of this can be done in Line 136.

---
It could be a really simple POC if you have the time for it. If not then I would like to contribute as much as I can.

Best Regards,
Deniz.

**Comment by DenizUgur** [comment]
Alright, so someone else already pointed this out to you in a Rectangle Pro [discussion](https://github.com/rxhanson/RectanglePro-Community/discussions/41). I'm not sure what's the timeline but I would've paid double the price of Pro if it even resembles FancyZone 😄

**Comment by rxhanson** [comment]
Sorry for the delay. 

It's still a ways off before I think it will make it into Rectangle Pro (and I won't make a promise - if I start down that path and it looks like something that I don't want to do then there's still a chance I might not do it, although I'm pretty sure I will add it in).

Definitely feel free to give it a go, it sounds like you're already down a path I would have suggested.

> 1. Read the configuration from a file or preferences. The option might describe 12x12 grid and each zone will be an area within this grid.

Preferences are handled in the Defaults class. If you're interested I can make a gist for my JsonDefault class that inherits from StringDefault and you can use that for managing prefs.

> 2. I see that you are defining some areas at Line 37. Maybe parse the configuration here. It would be enough to just snap to a location.

Sure, same concepts. Probably makes sense to make a "ZonesManager" or something like that and base it off of `SnappingManager`.

> 3. For the other option (multiple zones and bounding box of it), while left click + shift is pressed (or just right click) continue adding zones on to an array and find the bounding box of that array when the key combination is released. All of this can be done in Line 136.

I agree. I would still say it makes sense to create a "ZonesManager", and the `SnappingManager` could hold an instance of it and call it's methods for grabbing a `snapAreaContainingCursor`.
